### DISCUSSION TITLE
Certificate Authentication via IIS is painfully slow

### DISCUSSION THREAD

**Discussion by solarding** [dotnet/aspnetcore#25001] [General]
I tried my new API project with .netCore 3.1, where Client Certiticate Authentication is needed. 
According to official recommendations, the validation/trust check is done in IIS, and the authorization is done in my code   
OnCertificateValidated = context => {...}

I did exactly what ["Configure certificate authentication"](https://docs.microsoft.com/en-us/aspnet/core/security/authentication/certauth?view=aspnetcore-3.1) says, however the performance is a disaster.

The process of IIS cert. validation takes 10 to 15 secs. And the IIS does it every 1 or two minutes.
I see there is an option in the .net 5 previews called **AddCertificateCache().** 
**Does anymore know if this cache option is helpful? or it only affects cert. validation throug Kestrel** or also **IIS**? Should I ask the IIS team?

**Comment by solarding** [comment]


**Reply by Tratcher** [reply]
AddCertificateCache addresses perf issues in the ApsNetCore cert auth  handler, not IIS. You should try the latest 5.0 preview to see how much of an improvement it gets you.

Ideally you'd be able to turn off cert validation in IIS and do it all in the auth handler, but I don't know if that's possible.

10-15 seconds sounds like something is timing out. Have you tried disabling revocation checks?

**Reply by solarding** [reply]
Well that's probably what I need to do, to test the 5.0 preview.
I don't think it's a general problem in IIS because in other (older ASP.NET) web applications on the same server, where also SSL is enabled, there is no perf. problem. It is related to asp,net Core and the IIS bundle.

**Reply by Tratcher** [reply]
ASP.NET Core doesn't treat client certificates any differently at the IIS layer. The only addition is the middleware. If you remove the middleware do you still see an issue?

**Reply by solarding** [reply]
I do not understand where is the middleware that I can remove. 
The code I'm using is from MS Document **Configure certificate authentication in ASP.NET Core**
It is the eventhandler `OnCertificateValidate` got invoked very slow, comes in very late, it took a lot of time doing whatever it needs to do before calling this back 
```
services.AddAuthentication(
    CertificateAuthenticationDefaults.AuthenticationScheme)
    .AddCertificate(options =>
    {
        options.Events = new CertificateAuthenticationEvents
        {
            **OnCertificateValidated** = context =>
            {
                var validationService = context.HttpContext.RequestServices.GetService<ICertificateValidationService>();
                if (validationService.ValidateCertificate(context.ClientCertificate))
                {                   
                    context.Success();
                }           
                return Task.CompletedTask;
            }
        };
    });
```

**Reply by Tratcher** [reply]
I meant if you temporarily comment out all of the code you just showed there is the app still slow when it receives client certificates? I know the auth won't work without it, I just want you to test the server layer in isolation.

I'm assuming this is not an AspNetCore server issue, but an issue with AddCertificate. Note that AddCertificate's options include many settings that may affect the performance. You should try adjusting them to see which ones make it slow. Start with RevocationMode, that's a common performance bottleneck.

**Reply by solarding** [reply]
Thanks Tratcher, it is the RevocationMode of CertificaeAuthenticationOption, default "online" was the perf. killer.

**Reply by Tratcher** [reply]
That probably means the server doesn't have direct internet access and is timing out on that check. At least now you know what to investigate.
